system/security/plugins/jwtSecurity/jwtSecurity.cpp

```diff
@@ -0,0 +1,1149 @@
+            }
+
+            // Add file scope permissions in alphabetical (and implicitly, length) order
+            std::sort(fileScopePerms.begin(), fileScopePerms.end());
+            for (CollectedScopePerms::const_iterator x = fileScopePerms.begin(); x != fileScopePerms.end(); x++)
+            {
+                if (x->claim == "AllowFileScopeAccess")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Access, defaultFileAccess);
+                else if (x->claim == "AllowFileScopeRead")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Read, defaultFileAccess);
+                else if (x->claim == "AllowFileScopeWrite")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Write, defaultFileAccess);
+                else if (x->claim == "AllowFileScopeFull")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Full, defaultFileAccess);
+                else if (x->claim == "DenyFileScopeAccess")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Access, defaultFileAccess, true);
+                else if (x->claim == "DenyFileScopeRead")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Read, defaultFileAccess, true);
+                else if (x->claim == "DenyFileScopeWrite")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Write, defaultFileAccess, true);
+                else if (x->claim == "DenyFileScopeFull")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Full, defaultFileAccess, true);
+            }
+
+            gUserCache.set(subject, userInfo);
+
+            verificationResult = true;
+        }
+        catch (const std::runtime_error& e)
+        {
+            // Most jwt-cpp exceptions are subclasses of std::runtime_error
+            PROGLOG("CJwtSecurityManager::verifyToken error: %s", e.what());
+        }
+        catch (const std::logic_error& e)
+        {
+            PROGLOG("CJwtSecurityManager::logic_error: %s", e.what());
+        }
+        catch (const std::bad_cast& e)
+        {
+            PROGLOG("CJwtSecurityManager::bad_cast error: %s", e.what());
+        }
+        catch (...)
+        {
+            PROGLOG("CJwtSecurityManager::verifyToken unknown error");
+            throw;
+        }
+
+        return verificationResult;
+    }
+
+    /**
+     * Load the contents of the key if necessary.  Contents loaded when
+     * at least KEY_FILE_CHECK_INTERVAL have passed since last check time
+     *
+     * Function will throw an exception on error.
+     *
+     * @return  true
+     */
+    bool ensureKeyLoaded()
+    {
+        if ((keyContentsTimer.elapsedMs() > KEY_FILE_CHECK_INTERVAL_MS) || keyContents.empty())
+        {
+            try
+            {
+                StringBuffer    rawKeyContents;
+
+                getSecretValue(rawKeyContents, "system", secretsPath.c_str(), "key", true);
```

Comment:
@afishbeck what is the correct category for this type of secret?  Is system correct?  Does this plugin on run within esp, or elsewhere.  We currently have a difference in categories between secrets (storage, ecl, system) and vaults (storage, esp, ecl, ecl-user) which should be made consistent.

Reply:
@ghalliday The plugin runs in ESP and in DALI.  But @dcamper does DALI's use of the plugin make use of the secret?  System is meant for system level secrets but it will copy it to every pod.  I suppose we could add a category of secret just for use by security plugins since they are a special case?  It would minimize the pods that could read it.  If we agree we can add a category "security" and put them on ESP and DALI pods for now.

Btw, I've opened HPCC-24819 "Make secret and vault categories consistent".

Reply:
The secret is read only by the esp process.  Should we use a different category in this case?  And would this change affect the file system path used in bare-metal installations?

Reply:
If it is only read by esp, then using "esp" is probably correct (tony can confirm).  The parameter you have used for secretsPath is actually the name of the secret.
on baremetal (and container) secrets are mapped to opt/HPCCSystems/secrets/<category>/<secret-name>/<key>
e.g.
opt/HPCCSystems/secrets/esp/jwt-security-secret/key

Reply:
yes, using the ESP category would work.  In k8s the secret would only be exposed to the ESP pods.

Reply:
Changed to 'esp'